### ISSUE TITLE
A Deeper Look at Metafunctions

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#9774]
A Deeper Look at Metafunctions<br>
<br>
https://ift.tt/1ogMnhi<br>
<br>
<br>
<br>

<div><div><div><div>
<div>
<p>A Deeper Look at Metafunctions</p>
<p>by David Abrahams and Aleksey Gurtovoy</p>
<p>August 23, 2004</p>
<hr>
<div>
<p>Summary
</p>
<p>
This article explains how to define and use metafunctions, the compile-time equivalent of functions, and introduces the Boost Metaprogramming Library.
</p>
</div>
<p>
The Safe Bool Idiom </p>
<blockquote> <em>From Chapter 3 of <a href="http://www.awprofessional.com/titles/0321227255"></a></em>C++ Template Metaprogramming <em>by David Abrahams and Aleksey Gurtovoy. ISBN: 0321227255. Copyright (c) 2005 by Pearson Education, Inc. Reprinted with permission.</em> </blockquote> <p>With the foundation laid so far, we're ready to explore one of the most basic uses of template metaprogramming techniques: adding static type checking to traditionally unchecked operations. We'll look at a practical example from science and engineering that can find applications in almost any numerical code. Along the way you'll learn some important new concepts and get a taste of metaprogramming at a high level using the MPL.</p> <p>The first rule of doing physical calculations on paper is that the numbers being manipulated don't stand alone: most quantities have attached <em>dimensions</em>, to be ignored at our peril. As computations become more complex, keeping track of dimensions is what keeps us from inadvertently assigning a mass to what should be a length or adding acceleration to velocity—it establishes a type system for numbers.</p> <p>Manual checking of types is tedious, and as a result, it's also error-prone. When human beings become bored, their attention wanders and they tend to make mistakes. Doesn't type checking seem like the sort of job a computer might be good at, though? If we could establish a framework of C++ types for dimensions and quantities, we might be able to catch errors in formulae before they cause serious problems in the real world.</p> <p>Preventing quantities with different dimensions from interoperating isn't hard; we could simply represent dimensions as classes that only work with dimensions of the same type. What makes this problem interesting is that different dimensions <em>can</em> be combined, via multiplication or division, to produce arbitrarily complex new dimensions. For example, take Newton's law, which relates force to mass and acceleration:</p> <blockquote> <em>F</em> = <em>ma</em> </blockquote> <p>Since mass and acceleration have different dimensions, the dimensions of force must somehow capture their combination. In fact, the dimensions of acceleration are already just such a composite, a change in velocity over time:</p> <blockquote> <em>dv</em>/
<em>dt</em> </blockquote> <p>Since velocity is just change in distance (<em>l</em>) over time (<em>t</em>), the fundamental dimensions of acceleration are:</p> <blockquote>
(
<em>l</em>/
<em>t</em>)/
<em>t</em> = <em>l</em>/
<em>t</em>
<sup>2</sup>
</blockquote> <p>And indeed, acceleration is commonly measured in "meters per second squared." It follows that the dimensions of force must be:</p> <blockquote> <em>ml</em>/
<em>t</em>
<sup>2</sup>
</blockquote> <p>and force is commonly measured in kg(m/s<sup>2</sup>), or "kilogram-meters per second squared." When multiplying quantities of mass and acceleration, we multiply their dimensions as well and carry the result along, which helps us to ensure that the result is meaningful. The formal name for this bookkeeping is <strong>dimensional analysis</strong>, and our next task will be to implement its rules in the C++ type system. John Barton and Lee Nackman were the first to show how to do this in their seminal book, <em>Scientific and Engineering C++</em> <a href="https://www.artima.com/#bnack94">[BNack94]</a>. We will recast their approach here in metaprogramming terms.</p> <table>
<colgroup>
<col>
<col>
</colgroup>
<tbody><tr>
<td> <a href="https://www.artima.com/#id1">[BNack94]</a> </td> <td>John J. Barton and Lee R. Nackman, <em>Scientific and Engineering C++: an introduction with advanced techniques and examples</em> ISBN 0-201-53393-6, Addison Wesley, Reading Massachusetts, 1994</td> </tr></tbody>
</table>
<p>An international standard called <em>Système International d'Unites</em> (SI), breaks every quantity down into a combination of the dimensions <em>mass</em>, <em>length</em> (or <em>position</em>), <em>time</em>, <em>charge</em>, <em>temperature</em>, <em>intensity</em>, and <em>angle</em>. To be reasonably general, our system would have to be able to represent seven or more fundamental dimensions. It also needs the ability to represent composite dimensions which, like <em>force</em>, are built through multiplication or division of the fundamental ones.</p> <p>In general, a composite dimension is the product of powers of fundamental dimensions.<a href="https://www.artima.com/metafunctions3.html#notes">[1]</a> If we were going to represent these powers for manipulation at runtime, we could use an array of seven <code>int</code>s, with each position in the array holding the power of a different fundamental dimension:</p> <pre>
typedef int dimension[7]; // m  l  t  ...
dimension const mass      = {1, 0, 0, 0, 0, 0, 0};
dimension const length    = {0, 1, 0, 0, 0, 0, 0};
dimension const time      = {0, 0, 1, 0, 0, 0, 0};
...
</pre> <p>In that representation, force would be:</p> <pre>
dimension const force  = {1, 1, -2, 0, 0, 0, 0};
</pre> <p>that is, <em>mlt</em><sup>-2</sup>. However, if we want to get dimensions into the type system, these arrays won't do the trick: they're all the same type! Instead, we need types that <em>themselves</em> represent sequences of numbers, so that two masses have the same type and a mass is a different type from a length.</p> <p>Fortunately, the MPL provides us with a collection of <strong>type sequences</strong>. For example, we can build a sequence of the built-in signed integral types this way:</p> <pre>
#include &lt;boost/mpl/vector.hpp&gt;

typedef boost::mpl::vector&lt;
     signed char, short, int, long&gt; signed_types;
</pre> <p>How can we use a type sequence to represent numbers? Just as numerical metafunctions pass and return wrapper <em>types</em> having a nested <code>::value</code>, so numerical sequences are really sequences of wrapper types (another example of polymorphism). To make this sort of thing easier, MPL supplies the <code>int_&lt;N&gt;</code> class template, which presents its integral argument as a nested <code>::value</code>:</p> <pre>
#include &lt;boost/mpl/int.hpp&gt;

namespace mpl = boost::mpl; // namespace alias
static int const five = mpl::int_&lt;5&gt;::value;
</pre> <div> <p>
Namespace Aliases </p> <pre>
namespace <em>alias</em> = <em>namespace-name</em>;
</pre> <p>declares <em>alias</em> to be a synonym for <em>namespace-name</em>. Many examples in this book will use <code>mpl::</code> to indicate <code>boost::mpl::</code>, but will omit the alias that makes it legal C++.</p> </div> <p>In fact, the library contains a whole suite of integral constant wrappers such as <code>long_</code> and <code>bool_</code>, each one wrapping a different type of integral constant within a class template.</p> <p>Now we can build our fundamental dimensions:</p> <pre>
typedef mpl::vector&lt;
   mpl::int_&lt;1&gt;, mpl::int_&lt;0&gt;, mpl::int_&lt;0&gt;, mpl::int_&lt;0&gt;
 , mpl::int_&lt;0&gt;, mpl::int_&lt;0&gt;, mpl::int_&lt;0&gt;
&gt; mass;

typedef mpl::vector&lt;
   mpl::int_&lt;0&gt;, mpl::int_&lt;1&gt;, mpl::int_&lt;0&gt;, mpl::int_&lt;0&gt;
 , mpl::int_&lt;0&gt;, mpl::int_&lt;0&gt;, mpl::int_&lt;0&gt;
&gt; length;
...
</pre> <p>Whew! That's going to get tiring pretty quickly. Worse, it's hard to read and verify: The essential information, the powers of each fundamental dimension, is buried in repetitive syntactic "noise." Accordingly, MPL supplies <strong>integral sequence wrappers</strong> that allow us to write:</p> <pre>
#include &lt;boost/mpl/vector_c.hpp&gt;

typedef mpl::vector_c&lt;int,1,0,0,0,0,0,0&gt; mass;
typedef mpl::vector_c&lt;int,0,1,0,0,0,0,0&gt; length; // or position 
typedef mpl::vector_c&lt;int,0,0,1,0,0,0,0&gt; time;
typedef mpl::vector_c&lt;int,0,0,0,1,0,0,0&gt; charge;
typedef mpl::vector_c&lt;int,0,0,0,0,1,0,0&gt; temperature;
typedef mpl::vector_c&lt;int,0,0,0,0,0,1,0&gt; intensity;
typedef mpl::vector_c&lt;int,0,0,0,0,0,0,1&gt; angle;
</pre> <p>Even though they have different types, you can think of these <code>mpl::vector_c</code> specializations as being equivalent to the more verbose versions above that use <code>mpl::vector</code>.</p> <p>If we want, we can also define a few composite dimensions:</p> <pre>
// base dimension:        m l  t ... 
typedef mpl::vector_c&lt;int,0,1,-1,0,0,0,0&gt; velocity;     // l/t
typedef mpl::vector_c&lt;int,0,1,-2,0,0,0,0&gt; acceleration; // l/(t<sup>2</sup>)
typedef mpl::vector_c&lt;int,1,1,-1,0,0,0,0&gt; momentum;     // ml/t
typedef mpl::vector_c&lt;int,1,1,-2,0,0,0,0&gt; force;        // ml/(t<sup>2</sup>)
</pre> <p>And, incidentally, the dimensions of scalars (like pi) can be described as:</p> <pre>
typedef mpl::vector_c&lt;int,0,0,0,0,0,0,0&gt; scalar;
</pre> <p>The types listed above are still pure metadata; to typecheck real computations we'll need to somehow bind them to our runtime data. A simple numeric value wrapper, parameterized on the datatype <code>T</code> and on its dimensions, fits the bill:</p> <pre>
template &lt;class T, class Dimensions&gt;
struct quantity
{
    explicit quantity(T x)
       : m_value(x)
    {}

    T value() const { return m_value; }
 private:
    T m_value;
};
</pre> <p>Now we have a way to represent numbers associated with dimensions. For instance, we can say:</p> <pre>
quantity&lt;float,length&gt; l( 1.0f );
quantity&lt;float,mass&gt; m( 2.0f );
</pre> <p>Note that <code>Dimensions</code> doesn't appear anywhere in the definition of <code>quantity</code> outside the template parameter list; its <em>only</em> role is to ensure that <code>l</code> and <code>m</code> have different types. Because they do, we cannot make the mistake of assigning a length to a mass:</p> <pre>
m = l;    // compile-time type error
</pre> <p>We can now easily write the rules for addition and subtraction, since the dimensions of the arguments must always match.</p> <pre>
template &lt;class T, class D&gt;
quantity&lt;T,D&gt;
operator+(quantity&lt;T,D&gt; x, quantity&lt;T,D&gt; y)
{
  return quantity&lt;T,D&gt;(x.value() + y.value());
}

template &lt;class T, class D&gt;
quantity&lt;T,D&gt;
operator-(quantity&lt;T,D&gt; x, quantity&lt;T,D&gt; y)
{
  return quantity&lt;T,D&gt;(x.value() - y.value());
}
</pre> <p>These operators enable us to write code like:</p> <pre>
quantity&lt;float,<strong>length</strong>&gt; len1( 1.0f );
quantity&lt;float,<strong>length</strong>&gt; len2( 2.0f );

len1 = len1 + len2;   // OK
</pre> <p>but prevent us from trying to add incompatible dimensions:</p> <pre>
len1 = len2 + quantity&lt;float,<strong>mass</strong>&gt;( 3.7f ); // <strong>error</strong>
</pre> <p>Multiplication is a bit more complicated than addition and subtraction. So far, the dimensions of the arguments and results have all been identical, but when multiplying, the result will usually have different dimensions from either of the arguments. For multiplication, the relation:</p> <blockquote>
(
<em>x</em>
<sup>a</sup>)(
<em>x</em>
<sup>b</sup>) == <em>x</em> <sup>(a + b)</sup>
</blockquote> <p>implies that the exponents of the result dimensions should be the sum of corresponding exponents from the argument dimensions. Division is similar, except that the sum is replaced by a difference.</p> <p>To combine corresponding elements from two sequences, we'll use MPL's <code>transform</code> algorithm. <code>transform</code> is a metafunction that iterates through two input sequences in parallel, passing an element from each sequence to an arbitrary binary metafunction, and placing the result in an output sequence.</p> <pre>
template &lt;class Sequence1, class Sequence2, class BinaryOperation&gt;
struct transform;  // returns a Sequence
</pre> <p>The signature above should look familiar if you're acquainted with the STL <code>transform</code> algorithm that accepts two <em>runtime</em> sequences as inputs:</p> <pre>
template &lt;
    class InputIterator1, class InputIterator2
  , class OutputIterator, class BinaryOperation
&gt;
void transform(
    InputIterator1 start1, InputIterator2 finish1
  , InputIterator2 start2
  , OutputIterator result, BinaryOperation func);
</pre> <p>Now we just need to pass a <code>BinaryOperation</code> that adds or subtracts in order to multiply or divide dimensions with <code>mpl::transform</code>. If you look through the MPL reference manual, you'll come across <code>plus</code> and <code>minus</code> metafunctions that do just what you'd expect:</p> <pre>
#include &lt;boost/static_assert.hpp&gt;
#include &lt;boost/mpl/plus.hpp&gt;
#include &lt;boost/mpl/int.hpp&gt;
namespace mpl = boost::mpl;

BOOST_STATIC_ASSERT(( 
    mpl::plus&lt;
        mpl::int_&lt;2&gt;
      , mpl::int_&lt;3&gt;
    &gt;::type::value == 5
));
</pre> <div> <p> <code>BOOST_STATIC_ASSERT</code> </p> <p>is a macro that causes a compilation error if its argument is false. The double parentheses are required because the C++ preprocessor can't parse templates: it would otherwise be fooled by the comma into treating the condition as two separate macro arguments. Unlike its runtime analogue <code>assert(...)</code>, <code>BOOST_STATIC_ASSERT</code> can also be used at class scope, allowing us to put assertions in our metafunctions. See Chapter 8 for an in-depth discussion.</p> </div> <p>At this point it might seem as though we have a solution, but we're not quite there yet. A naive attempt to apply the <code>transform</code> algorithm in the implementation of <code>operator*</code> yields a compiler error:</p> <pre>
#include &lt;boost/mpl/transform.hpp&gt;

template &lt;class T, class D1, class D2&gt;
quantity&lt; 
    T
  , typename mpl::transform&lt;D1,D2,mpl::plus&gt;::type
&gt;
operator*(quantity&lt;T,D1&gt; x, quantity&lt;T,D2&gt; y) { ... }
</pre> <p>It fails because the protocol says that metafunction arguments must be types, and <code>plus</code> is not a type, but a class template. Somehow we need to make metafunctions like <code>plus</code> fit the metadata mold.</p> <p>One natural way to introduce polymorphism between metafunctions and metadata is to employ the wrapper idiom that gave us polymorphism between types and integral constants. Instead of a nested integral constant, we can use a class template nested within a <strong>metafunction class</strong>:</p> <pre>
struct plus_f
{
    template &lt;class T1, class T2&gt;
    struct apply
    {
       typedef typename mpl::plus&lt;T1,T2&gt;::type type;
    };
};
</pre> <div> <p>
Definition </p> <p>A <strong>Metafunction Class</strong> is a class with a publicly accessible nested metafunction called <code>apply</code>.</p> </div> <p>Whereas a metafunction is a template but not a type, a metafunction class wraps that template within an ordinary non-templated class, which <em>is</em> a type. Since metafunctions operate on and return types, a metafunction class can be passed as an argument to, or returned from, another metafunction.</p> <p>Finally, we have a <code>BinaryOperation</code> type that we can pass to <code>transform</code> without causing a compilation error:</p> <pre>
template &lt;class T, class D1, class D2&gt;
quantity&lt; 
    T
  , typename mpl::transform&lt;D1,D2,<strong>plus_f</strong>&gt;::type  // new dimensions
&gt;
operator*(quantity&lt;T,D1&gt; x, quantity&lt;T,D2&gt; y)
{
    typedef typename mpl::transform&lt;D1,D2,<strong>plus_f</strong>&gt;::type dim;
    return quantity&lt;T,dim&gt;( x.value() * y.value() );
}
</pre> <p>Now, if we want to compute the force exterted by gravity on a 5 kilogram laptop computer, that's just the acceleration due to gravity (9.8 m/sec<sup>2</sup>) times the mass of the laptop:</p> <pre>
quantity&lt;float,mass&gt; m(5.0f);
quantity&lt;float,acceleration&gt; a(9.8f);
std::cout &lt;&lt; "force = " &lt;&lt; (m * a).value();
</pre> <p>Our <code>operator*</code> multiplies the runtime values (resulting in 6.0f), and our metaprogram code uses <code>transform</code> to sum the meta-sequences of fundamental dimension exponents, so that the result type contains a representation of a new list of exponents, something like:</p> <pre>
vector_c&lt;int,1,1,-2,0,0,0,0&gt;
</pre> <p>However, if we try to write:</p> <pre>
quantity&lt;float,force&gt; f = m * a;
</pre> <p>we'll run into a little problem. Although the result of <code>m</code> <span>*</span> <span>a</span> does indeed represent a force with exponents of mass, length, and time 1, 1, and -2 respectively, the type returned by <code>transform</code> isn't a specialization of <code>vector_c</code>. Instead, <code>transform</code> works generically on the elements of its inputs and builds a new sequence with the appropriate elements: a type with many of the same sequence properties as <code>vector_c&lt;int,1,1,-2,0,0,0,0&gt;</code>, but with a different C++ type altogether. If you want to see the type's full name, you can try to compile the example yourself and look at the error message, but the exact details aren't important. The point is that <code>force</code> names a different type, so the assignment above will fail.</p> <p>In order to resolve the problem, we can add an implicit conversion from the multiplication's result type to <code>quantity&lt;float,force&gt;</code>. Since we can't predict the exact types of the dimensions involved in any computation, this conversion will have to be templated, something like:</p> <pre>
template &lt;class T, class Dimensions&gt;
struct quantity
{
    // converting constructor
    template &lt;class OtherDimensions&gt;
    quantity(quantity&lt;T,OtherDimensions&gt; const&amp; rhs)
      : m_value(rhs.value())
    {
    }
    ...
</pre> <p>Unfortunately, such a general conversion undermines our whole purpose, allowing nonsense such as:</p> <pre>
// Should yield a force, not a mass!
quantity&lt;float,mass&gt; bogus = m * a;
</pre> <p>Luckily, we can correct that problem using another MPL algorithm, <code>equal</code>, which tests that two sequences have the same elements:</p> <pre>
template &lt;class OtherDimensions&gt;
quantity(quantity&lt;T,OtherDimensions&gt; const&amp; rhs)
  : m_value(rhs.value())
{
    BOOST_STATIC_ASSERT((
       mpl::equal&lt;Dimensions,OtherDimensions&gt;::type::value
    ));
}
</pre> <p>Now, if the dimensions of the two quantities fail to match, the assertion will cause a compilation error.</p> <p>Division is similar to multiplication, but instead of adding exponents, we must subtract them. Rather than writing out a near duplicate of <code>plus_f</code>, we can use the following trick to make <code>minus_f</code> much simpler:</p> <pre>
struct minus_f
{
    template &lt;class T1, class T2&gt;
    struct apply
      : mpl::minus&lt;T1,T2&gt; {};
};
</pre> <p>Here <code>minus_f::apply</code> uses inheritance to expose the nested <code>type</code> of its base class, <code>mpl::minus</code>, so we don't have to write:</p> <pre>
typedef typename ...::type type
</pre> <p>We don't have to write <code>typename</code> here (in fact, it would be illegal), because the compiler knows that dependent names in <code>apply</code>'s initializer list must be base classes.<a href="https://www.artima.com/metafunctions3.html#notes">[2]</a> This powerful simplification is known as <strong>metafunction forwarding</strong>; we'll apply it often as the book goes on.<a href="https://www.artima.com/metafunctions3.html#notes">[3]</a></p> <p>Syntactic tricks notwithstanding, writing trivial classes to wrap existing metafunctions is going to get boring pretty quickly. Even though the definition of <code>minus_f</code> was far less verbose than that of <code>plus_f</code>, it's still an awful lot to type. Fortunately, MPL gives us a <em>much</em> simpler way to pass metafunctions around. Instead of building a whole metafunction class, we can invoke <code>transform</code> this way:</p> <pre>
typename mpl::transform&lt;D1,D2, <strong>mpl::minus&lt;_1,_2&gt;</strong> &gt;::type
</pre> <p>Those funny looking arguments (<code>_1</code> and <code>_2</code>) are known as <strong>placeholders</strong>, and they signify that when the <code>transform</code>'s <code>BinaryOperation</code> is invoked, its first and second arguments will be passed on to <code>minus</code> in the positions indicated by <code>_1</code> and <code>_2</code>, respectively. The whole type <code>mpl::minus&lt;_1,_2&gt;</code> is known as a <strong>placeholder expression</strong>.</p> <div> <p>
Note </p> <p>MPL's placeholders are in the <code>mpl::placeholders</code> namespace and defined in <code>boost/mpl/placeholders.hpp</code>. In this book we will usually assume that you have written:</p> <pre>
#include&lt;boost/mpl/placeholders.hpp&gt;
using namespace mpl::placeholders;
</pre> <p>so that they can be accessed without qualification.</p> </div> <p>Here's our division operator written using placeholder expressions:</p> <pre>
template &lt;class T, class D1, class D2&gt;
quantity&lt; 
    T
  , typename mpl::transform&lt;D1,D2,<strong>mpl::minus&lt;_1,_2&gt;</strong> &gt;::type
&gt;
operator/(quantity&lt;T,D1&gt; x, quantity&lt;T,D2&gt; y)
{
   typedef typename 
     mpl::transform&lt;D1,D2,<strong>mpl::minus&lt;_1,_2&gt;</strong> &gt;::type dim;

   return quantity&lt;T,dim&gt;( x.value() / y.value() );
}
</pre> <p>This code is considerably simpler. We can simplify it even further by factoring the code that calculates the new dimensions into its own metafunction:</p> <pre>
template &lt;class D1, class D2&gt;
struct <strong>divide_dimensions</strong>
  : mpl::transform&lt;D1,D2,mpl::minus&lt;_1,_2&gt; &gt; // forwarding again
{};

template &lt;class T, class D1, class D2&gt;
quantity&lt;T, typename <strong>divide_dimensions&lt;D1,D2&gt;</strong>::type&gt;
operator/(quantity&lt;T,D1&gt; x, quantity&lt;T,D2&gt; y)
{
   return quantity&lt;T, typename <strong>divide_dimensions&lt;D1,D2&gt;</strong>::type&gt;(
      x.value() / y.value());
}
</pre> <p>Now we can verify our "force-on-a-laptop" computation by reversing it, as follows:</p> <pre>
quantity&lt;float,mass&gt; m2 = f/a;
float rounding_error = std::abs((m2 - m).value());
</pre> <p>If we got everything right, <code>rounding_error</code> should be very close to zero. These are boring calculations, but they're just the sort of thing that could ruin a whole program (or worse) if you got them wrong. If we had written <code>a/f</code> instead of <code>f/a</code>, there would have been a compilation error, preventing a mistake from propagating throughout our program.</p> <p>In the previous section we used two different forms—metafunction classes and placeholder expressions—to pass and return metafunctions just like any other metadata. Bundling metafunctions into "first class metadata" allows <code>transform</code> to perform an infinite variety of different operations: in our case, multiplication and division of dimensions. Though the idea of using functions to manipulate other functions may seem simple, its great power and flexibility <a href="https://www.artima.com/#hudak89">[Hudak89]</a> has earned it a fancy title: <strong>higher-order functional programming</strong>. A function that operates on another function is known as a <strong>higher-order function</strong>. It follows that <code>transform</code> is a higher-order metafunction: a metafunction that operates on another metafunction.</p> <table>
<colgroup>
<col>
<col>
</colgroup>
<tbody><tr>
<td> <a href="https://www.artima.com/#id5">[Hudak89]</a> </td> <td>Paul Hudak, "Conception, evolution, and application of functional programming languages," ACM Press New York, NY, USA Pages: 359 - 411, 1989 ISSN:0360-0300 https://ift.tt/Q0KdAzn</td> </tr></tbody>
</table>
<p>Now that we've seen the power of higher-order metafunctions at work, it would be good to be able to create new ones. In order to explore the basic mechanisms, let's try a simple example. Our task is to write a metafunction called <code>twice</code>, which—given a unary metafunction <em>f</em> and arbitrary metadata <em>x</em>—computes:</p> <blockquote> <em>twice</em>(
<em>f</em>, <em>x</em>) := <em>f</em>(
<em>f</em>(
<em>x</em>)) </blockquote> <p>This might seem like a trivial example, and in fact it is. You won't find much use for <code>twice</code> in real code. We hope you'll bear with us anyway: Because it doesn't do much more than accept and invoke a metafunction, <code>twice</code> captures all the essential elements of "higher-orderness" without any distracting details.</p> <p>If <em>f</em> is a metafunction class, the definition of <code>twice</code> is straightforward:</p> <pre>
template &lt;class F, class X&gt;
struct twice
{
    typedef typename F::template apply&lt;X&gt;::type once;    // f(x)
    typedef typename F::template apply&lt;once&gt;::type type; // f(f(x))
};
</pre> <p>Or, applying metafunction forwarding:</p> <pre>
template &lt;class F, class X&gt;
struct twice
  : F::template apply&lt;
       typename F::template apply&lt;X&gt;::type
    &gt;
{};
</pre> <div> <p>
C++ Language Note </p> <p>The C++ standard requires the <code>template</code> keyword when we use a <strong>dependent name</strong> that refers to a member template. <code>F::apply</code> may or may not name a template, <em>depending</em> on the particular <code>F</code> that is passed. See Appendix B for more information about <code>template</code>.</p> </div> <p>Given the need to sprinkle our code with the <code>template</code> keyword, it would be nice to reduce the syntactic burden of invoking metafunction classes. As usual, the solution is to factor the pattern into a metafunction:</p> <pre>
template &lt;class UnaryMetaFunctionClass, class Arg&gt;
struct apply1
  : UnaryMetaFunctionClass::template apply&lt;Arg&gt;
{};
</pre> <p>Now <code>twice</code> is just:</p> <pre>
template &lt;class F, class X&gt;
struct twice
  : apply1&lt;F, typename apply1&lt;F,X&gt;::type&gt;
{};
</pre> <p>To see <code>twice</code> at work, we can apply it to a little metafunction class built around the <code>add_pointer</code> metafunction:</p> <pre>
struct add_pointer_f
{
    template &lt;class T&gt;
    struct apply : boost::add_pointer&lt;T&gt; {};
};
</pre> <p>Now we can use <code>twice</code> with <code>add_pointer_f</code> to build pointers-to-pointers:</p> <pre>
BOOST_STATIC_ASSERT((
    boost::is_same&lt;
         twice&lt;add_pointer_f, int&gt;::type
       , int**
    &gt;::value
));
</pre> <p>Our implementation of <code>twice</code> already works with metafunction classes. Ideally, we would like it to work with placeholder expressions too, much the same as <code>mpl::transform</code> allows us to pass either form. For example, we would like to be able to write:</p> <pre>
template &lt;class X&gt;
struct two_pointers
    : twice&lt;<strong>boost::add_pointer&lt;_1&gt;</strong>, X&gt;
{};
</pre> <p>But when we look at the implementation of <code>boost::add_pointer</code>, it becomes clear that the current definition of <code>twice</code> can't work that way.</p> <pre>
template &lt;class T&gt;
struct add_pointer
{
    typedef T* type;
};
</pre> <p>To be invokable by <code>twice</code>, <code>boost::add_pointer&lt;_1&gt;</code> would have to be a metafunction class, along the lines of <code>add_pointer_f</code>. Instead, it's just a nullary metafunction returning the almost senseless type <code>_1*</code>. Any attempt to use <code>two_pointers</code> will fail when <code>apply1</code> reaches for a nested <code>::apply</code> metafunction in <code>boost::add_pointer&lt;_1&gt;</code> and finds that it doesn't exist.</p> <p>We've determined that we don't get the behavior we want automatically, so what next? Since <code>mpl::transform</code> can do this sort of thing, there ought to be a way for us to do it too—and so there is.</p> <p>We can <em>generate</em> a metafunction class from <code>boost::add_pointer&lt;_1&gt;</code>, using MPL's <code>lambda</code> metafunction:</p> <pre>
template &lt;class X&gt;
struct two_pointers
  : twice&lt;<strong>typename mpl::lambda&lt;boost::add_pointer&lt;_1&gt; &gt;::type</strong>, X&gt;
{};

BOOST_STATIC_ASSERT((
    boost::is_same&lt;
        typename two_pointers&lt;int&gt;::type
      , int**
    &gt;::value
));
</pre> <p>We'll refer to metafunction classes like <code>add_pointer_f</code> and placeholder expressions like <code>boost::add_pointer&lt;_1&gt;</code> as <strong>lambda expressions</strong>. The term, meaning "unnamed function object," was introduced in the 1930s by the logician Alonzo Church as part of a fundamental theory of computation he called the <em>lambda-calculus</em>.<a href="https://www.artima.com/metafunctions3.html#notes">[4]</a> MPL uses the somewhat obscure word <code>lambda</code> because of its well-established precedent in functional programming languages.</p> <p>Although its primary purpose is to turn placeholder expressions into metafunction classes, <code>mpl::lambda</code> can accept any lambda expression, even if it's already a metafunction class. In that case, <code>lambda</code> returns its argument unchanged. MPL algorithms like <code>transform</code> call <code>lambda</code> internally, before invoking the resulting metafunction class, so that they work equally well with either kind of lambda expression. We can apply the same strategy to <code>twice</code>:</p> <pre>
template &lt;class F, class X&gt;
struct twice
   : apply1&lt;
         typename mpl::lambda&lt;F&gt;::type
       , typename apply1&lt;
             typename mpl::lambda&lt;F&gt;::type
           , X
         &gt;::type
     &gt;
{};
</pre> <p>Now we can use <code>twice</code> with metafunction classes <em>and</em> placeholder expressions:</p> <pre>
int* x;

twice&lt;<strong>add_pointer_f</strong>, int&gt;::type           p = &amp;x;
twice&lt;<strong>boost::add_pointer&lt;_1&gt;</strong>, int&gt;::type q = &amp;x;
</pre> <p>Invoking the result of <code>lambda</code> is such a common pattern that MPL provides an <code>apply</code> metafunction to do just that. Using <code>mpl::apply</code>, our flexible version of <code>twice</code> becomes:</p> <pre>
#include &lt;boost/mpl/apply.hpp&gt;

template &lt;class F, class X&gt;
struct twice
   : mpl::apply&lt;F, typename mpl::apply&lt;F,X&gt;::type&gt;
{};
</pre> <p>You can think of <code>mpl::apply</code> as being just like the <code>apply1</code> template that we wrote, with two additional features:</p> <ol>
<li> <p>While <code>apply1</code> operates only on metafunction classes, the first argument to <code>mpl::apply</code> can be any lambda expression (including those built with placeholders).</p> </li> <li> <p>While <code>apply1</code> accepts only one additional argument to which the metafunction class will be applied, <code>mpl::apply</code> can invoke its first argument on any number from zero to five additional arguments.<a href="https://www.artima.com/metafunctions3.html#notes">[5]</a> For example:</p> <pre>
// binary lambda expression applied to 2 additional arguments
mpl::apply&lt;
    mpl::plus&lt;_1,_2&gt;   
  , <strong>mpl::int_&lt;6&gt;</strong>
  , <strong>mpl::int_&lt;7&gt;</strong>
&gt;::type::value // == 13
</pre> </li> </ol>
<div> <p>
Guideline </p> <p>When writing a metafunction that invokes one of its arguments, use <code>mpl::apply</code> so that it works with lambda expressions.</p> </div> <p>Lambda expressions provide much more than just the ability to pass a metafunction as an argument. The two capabilities described next combine to make lambda expressions an invaluable part of almost every metaprogramming task.</p> <p>Consider the lambda expression <code>mpl::plus&lt;_1,_1&gt;</code>. A single argument is directed to both of <code>plus</code>'s parameters, thereby adding a number to itself. Thus, a <em>binary</em> metafunction, <code>plus</code>, is used to build a <em>unary</em> lambda expression. In other words, we've created a whole new computation! We're not done yet, though: By supplying a non-placeholder as one of the arguments, we can build a unary lambda expression that adds a fixed value, say 42, to its argument:</p> <pre>
mpl::plus&lt;_1, mpl::int_&lt;42&gt; &gt;
</pre> <p>The process of binding argument values to a subset of a function's parameters is known in the world of functional programming as <strong>partial function application</strong>.</p> <p>Lambda expressions can also be used to assemble more interesting computations from simple metafunctions. For example, the following expression, which multiplies the sum of two numbers by their difference, is a <strong>composition</strong> of the three metafunctions <code>multiplies</code>, <code>plus</code>, and <code>minus</code>:</p> <pre>
mpl::multiplies&lt;mpl::plus&lt;_1,_2&gt;, mpl::minus&lt;_1,_2&gt; &gt;
</pre> <p>When evaluating a lambda expression, MPL checks to see if any of its arguments are themselves lambda expressions, and evaluates each one that it finds. The results of these inner evaluations are substituted into the outer expression before it is evaluated.</p> <p>Now that you have an idea of the semantics of MPL's <code>lambda</code> facility, let's formalize that understanding and look at things a little more deeply.</p> <p>The definition of "placeholder" may surprise you:</p> <div> <p>
Definition </p> <p>A <strong>placeholder</strong> is a metafunction class of the form <code>mpl::arg&lt;X&gt;</code>.</p> </div> <p>The convenient names <code>_1</code>, <code>_2</code>,... <code>_5</code> are actually <code>typedef</code>s for specializations of <code>mpl::arg</code> that simply select the <em>N</em>th argument for any <em>N</em>. <a href="https://www.artima.com/metafunctions3.html#notes">[6]</a> The implementation of placeholders looks something like this:</p> <pre>
namespace boost { namespace mpl { namespace placeholders {

template &lt;int N&gt; struct arg; // forward declarations
struct void_;

template &lt;&gt;
struct arg&lt;<strong>1</strong>&gt;
{
    template &lt;
      class <strong>A1</strong>, class A2 = void_, ... class A<em>m</em> = void_&gt;
    struct apply
    {
        typedef <strong>A1</strong> type; // return the first argument
    };
};
typedef <strong>arg&lt;1&gt; _1</strong>;

template &lt;&gt;
struct arg&lt;<strong>2</strong>&gt;
{
    template &lt;
      class A1, class <strong>A2</strong>, class A3 = void_, ...class A<em>m</em> = void_
    &gt;
    struct apply
    {
        typedef <strong>A2</strong> type; // return the second argument
    };
};
typedef <strong>arg&lt;2&gt; _2</strong>;

<em>more specializations and typedefs...</em>

}}}
</pre> <p>Remember that invoking a metafunction class is the same as invoking its nested <code>apply</code> metafunction. When a placeholder in a lambda expression is evaluated, it is invoked on the expression's actual arguments, returning just one of them. The results are then substituted back into the lambda expression and the evaluation process continues.</p> <p>There's one special placeholder, known as the <strong>unnamed placeholder</strong>, that we haven't yet defined:</p> <pre>
namespace boost { namespace mpl { namespace placeholders {

<strong>typedef arg&lt;-1&gt; _;</strong> // the unnamed placeholder

}}}
</pre> <p>The details of its implementation aren't important; all you really need to know about the unnamed placeholder is that it gets special treatment. When a lambda expression is being transformed into a metafunction class by <code>mpl::lambda</code>,</p> <blockquote>
the <em>n</em>th appearance of the unnamed placeholder <em>in a given template specialization</em> is replaced with <code>_</code>
<em>n</em>. </blockquote> <p>So, for example, every row of Table 3.1 below contains two equivalent lambda expressions.</p> <table>
<caption>
Unnamed Placeholder Semantics </caption> <colgroup>
<col>
<col>
</colgroup>
<tbody>
<tr>
<td> <pre>
mpl::plus&lt;_,_&gt;
</pre> </td> <td> <pre>
mpl::plus&lt;_1,_2&gt;
</pre> </td> </tr>
<tr>
<td> <pre>
boost::is_same&lt;
    _
  , boost::add_pointer&lt;_&gt;
&gt;
</pre> </td> <td> <pre>
boost::is_same&lt;
    _1
  , boost::add_pointer&lt;_1&gt;
&gt;
</pre> </td> </tr>
<tr>
<td> <pre>
mpl::multiplies&lt;
   mpl::plus&lt;_,_&gt;
 , mpl::minus&lt;_,_&gt;
&gt;
</pre> </td> <td> <pre>
mpl::multiplies&lt;
   mpl::plus&lt;_1,_2&gt;
 , mpl::minus&lt;_1,_2&gt;
&gt;
</pre> </td> </tr>
</tbody>
</table>
<p>Especially when used in simple lambda expressions, the unnamed placeholder often eliminates just enough syntactic "noise" to significantly improve readability.</p> <p>Now that you know just what <em>placeholder</em> means, we can define <em>placeholder expression</em>:</p> <div> <p>
Definition </p> <p>A placeholder expression is either:</p> <blockquote> <blockquote> </blockquote> <p><em>or</em></p> <blockquote> <ul>
<li>a template specialization with at least one argument that is a placeholder expression.</li> </ul>
</blockquote> </blockquote> </div> <p>In other words, a placeholder expression always involves a placeholder.</p> <p>There is just one detail of placeholder expressions that we haven't discussed yet. MPL uses a special rule to make it easier to integrate ordinary templates into metaprograms: After all of the placeholders have been replaced with actual arguments, if the resulting template specialization <em>X</em> doesn't have a nested <code>::type</code>, the result is just <em>X</em> itself.</p> <p>For example, <code>mpl::apply&lt;std::vector&lt;_&gt;,</code> <span>T&gt;</span> is always just <code>std::vector&lt;T&gt;</code>. If it weren't for this behavior, we would have to build trivial metafunctions to create ordinary template specializations in lambda expressions:</p> <pre>
// trivial std::vector generator
template&lt;class U&gt; 
struct make_vector { typedef std::vector&lt;U&gt; type; };

typedef mpl::apply&lt;<strong>make_vector&lt;_&gt;</strong>, T&gt;::type vector_of_t;
</pre> <p>Instead, we can simply write:</p> <pre>
typedef mpl::apply&lt;<strong>std::vector&lt;_&gt;</strong>, T&gt;::type vector_of_t;
</pre> <p>Recall the definition of <code>always_int</code> from the previous chapter:</p> <pre>
struct always_int
{
    typedef int type;
};
</pre> <p>Nullary metafunctions might not seem very important at first, since something like <code>add_pointer&lt;int&gt;</code> could be replaced by <code>int*</code> in any lambda expression where it appears. Not all nullary metafunctions are that simple, though:</p> <pre>
struct add_pointer_f
{
    template &lt;class T&gt;
    struct apply : boost::add_pointer&lt;T&gt; {};
};
typedef mpl::vector&lt;int, char*, double&amp;&gt; seq;
typedef <strong>mpl::transform&lt;seq, add_pointer_f&gt;</strong> calc_ptr_seq;
</pre> <p>Note that <code>calc_ptr_seq</code> is a nullary metafunction, since it has <code>transform</code>'s nested <code>::type</code>. A C++ template is not instantiated until we actually "look inside it," though. Just naming <code>calc_ptr_seq</code> does not cause it to be evaluated, since we haven't accessed its <code>::type</code> yet.</p> <p>Metafunctions can be invoked <em>lazily</em>, rather than immediately upon supplying all of their arguments. We can use <strong>lazy evaluation</strong> to improve compilation time when a metafunction result is only going to be used conditionally. We can sometimes also avoid contorting program structure by <em>naming</em> an invalid computation without actually performing it. That's what we've done with <code>calc_ptr_seq</code> above, since you can't legally form <code>double&amp;*</code>. Laziness and all of its virtues will be a recurring theme throughout this book.</p> <p>By now you should have a fairly complete view of the fundamental concepts and language of both template metaprogramming in general and of the Boost Metaprogramming Library. This section reviews the highlights.</p> <dl>
<dt>
<strong>Metafunction forwarding.</strong>
</dt> <dd>
The technique of using public derivation to supply the nested <code>type</code> of a metafunction by accessing the one provided by its base class.
</dd> <dt>
<strong>Metafunction class.</strong>
</dt> <dd>
The most basic way to formulate a compile-time function so that it can be treated as polymorphic metadata; that is, as a type. A metafunction class is a class with a nested metafunction called <code>apply</code>.
</dd> <dt>
<strong>MPL.</strong>
</dt> <dd>
Most of this book's examples will use the Boost Metaprogramming Library. Like the Boost type traits headers, MPL headers follow a simple convention: <pre>
#include &lt;boost/mpl/<em>component-name</em>.hpp&gt;
</pre> <p>If the component's name ends in an underscore, however, the corresponding MPL header name does not include the trailing underscore. For example, <code>mpl::bool_</code> can be found in <code>&lt;boost/mpl/bool.hpp&gt;</code>. Where the library deviates from this convention, we'll be sure to point it out to you.</p> </dd> </dl>
<dl>
<dt>
<strong>Higher-order function.</strong>
</dt> <dd>
A function that operates on or returns a function. Making metafunctions polymorphic with other metadata is a key ingredient in higher-order metaprogramming.
</dd> <dt>
<strong>Lambda expression.</strong>
</dt> <dd>
Simply put, a lambda expression is callable metadata. Without some form of callable metadata, higher-order metafunctions would be impossible. Lambda expressions have two basic forms: <em>metafunction classes</em> and <em>placeholder expressions</em>.
</dd> <dt>
<strong>Placeholder expression.</strong>
</dt> <dd> <p>A kind of lambda expression that, through the use of placeholders, enables in-place <em>partial metafunction application</em> and <em>metafunction composition</em>. As you will see throughout this book, these features give us the truly amazing ability to build up almost any kind of complex type computation from more primitive metafunctions, right at its point of use:</p> <pre>
// find the position of a type x in some_sequence such that:
//         x is convertible to 'int'
//      &amp;&amp; x is not 'char'
//      &amp;&amp; x is not a floating type
typedef mpl::find_if&lt;
      some_sequence
    , mpl::and_&lt;
          boost::is_convertible&lt;_1,int&gt;
        , mpl::not_&lt;boost::is_same&lt;_1,char&gt; &gt;
        , mpl::not_&lt;boost::is_float&lt;_1&gt; &gt;
      &gt;
    &gt;::type iter;
</pre> <p>Placeholder expressions make good on the promise of algorithm reuse without forcing us to write new metafunction classes. The corresponding capability is often sorely missed in the runtime world of the STL, since it is often much easier to write a loop by hand than it is to use standard algorithms, despite their correctness and efficiency advantages.</p> </dd> </dl>
<dl>
<dt>
<strong>The <code>lambda</code> metafunction.</strong>
</dt> <dd>
A metafunction that transforms a lambda expression into a corresponding metafunction class. For detailed information on <code>lambda</code> and the lambda evaluation process, please see the MPL reference manual.
</dd> <dt>
<strong>The <code>apply</code> metafunction.</strong>
</dt> <dd>
A metafunction that invokes its first argument, which must be a lambda expression, on its remaining arguments. In general, to invoke a lambda expression, you should always pass it to <code>mpl::apply</code> along with the arguments you want to apply it to in lieu of using <code>lambda</code> and invoking the result "manually."
</dd> <dt>
<strong>Lazy evaluation.</strong>
</dt> <dd>
A strategy of delaying computation until its result is required, thereby avoiding any unneccessary computation and any associated unneccessary errors. Metafunctions are only invoked when we access their nested <code>::type</code>s, so we can supply all of their arguments without performing any computation and delay evaluation to the last possible moment.
</dd> </dl>
<h1>Notes</h1>
<a></a> <ol>
<li>Divisors just contribute negative exponents, since 1/<em>x</em> = <em>x</em><sup>-1</sup>.</li> <li>In case you're wondering, the same approach could have been applied to <code>plus_f</code>, but since it's a little subtle, we introduced the straightforward but verbose formulation first.</li> <li>Users of EDG-based compilers should consult Appendix C for a caveat about metafunction forwarding. You can tell whether you have an EDG compiler by checking the preprocessor symbol <code>__EDG_VERSION__</code>, which is defined by all EDG-based compilers.</li> <li>See <a href="http://en.wikipedia.org/wiki/Lambda_calculus"> http://en.wikipedia.org/wiki/Lambda_calculus</a> for an in-depth treatment, including a reference to Church's paper proving that the equivalence of lambda expressions is in general not decidable. </li> <li>See the Configuration Macros section of the MPL reference manual for a description of how to change the maximum number of arguments handled by <code>mpl::apply</code>.</li> <li>MPL provides five placeholders by default. See the Configuration Macros section of MPL reference manual for a description of how to change the number of placeholders provided.</li> </ol>
<a> <h1>Resources</h1></a>
<p></p> <p><a href="http://www.amazon.com/exec/obidos/ASIN/0321227255/billvennersA/"> <img align="left" src="https://www.artima.com/images/cppmeta.jpg"></a> David Abrahams and Aleksey Gurtovoy are the authors of <em>Template Metaprogramming</em>, which can be pre-ordered from on Amazon.com at:<br><a href="http://www.amazon.com/exec/obidos/ASIN/0321227255/billvennersA/"> http://www.amazon.com/exec/obidos/ASIN/0321227255/</a><br></p>
<h4>Talk back!</h4>
<p>
Have an opinion? Readers have already posted <a href="https://www.artima.com/articles/a-deeper-look-at-metafunctions/flat-comments">9 comments</a> about this article. Why not <a href="https://www.artima.com/articles/a-deeper-look-at-metafunctions/flat-comments">add yours</a>?
</p>
<h4>About the authors</h4>
<p>
</p>
<p>Nigel Warren is cofounder and director of technology at
IntaMission Ltd., where he researches and designs agile and
evolvable software infrastructures for next-generation
distributed systems.</p>
<p>Philip and Nigel are also the joint authors of <em>JavaSpaces
in Practice</em> and <em>Java in Practice,</em> both published by
Addison-Wesley.
</p>
<p>
</p>
<p><strong>Aleksey Gurtovoy</strong> is a technical lead at
MetaCommunications, Inc, and a contributing member of the C++
Boost community. He holds a MS degree in Computer Science from
Krasnoyarsk Technical State University, Russia. He can be reached
at agurtovoy@meta-comm.com.</p>
</div>
</div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via artima - Home <br>
<br>
September 2, 2023 at 09:52AM
